Create initial draft of Implementation Guide for NMOS Controllers

rmsporter opened the pull request:
Initial draft of Implementation Guide for NMOS Controllers, focusing only on IS-04, IS-05 and BCP-003-01 so far, but laying out the structure for sections relating to all NMOS specs.

rmsporter committed:
Create initial draft of Implementation Guide for NMOS Controllers

Initial draft of Implementation Guide for NMOS Controllers, focusing only on IS-04 so far, but laying out the structure for sections relating to all NMOS specs.

rmsporter committed:
Update Controller Guide.md

Added description of role of Controllers in IS-05 section.

peterbrightwell committed:
Fix lint

rmsporter committed:
Add section on BCP-003-01 to Controller Guide.md

rmsporter committed:
Add IS-04 and IS-05 diagrams and modify existing roles diagram

rmsporter committed:
Fix excessive whitespace in diagrams

rmsporter committed:
Update Controller Guide.md

Add working placeholder link for BCP-003-01 Controllers section

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+## Scope
+This document is intended as a guide for implementers or users of Controllers within NMOS-enabled networked media systems. The document defines what a Controller is and outlines the requirements of a Controller with respect to each existing NMOS specification through references to the relevant sections of those documents.
+
+The document focuses primarily on guidance for the core NMOS specifications of IS-04, IS-05 and BCP-003-01. However, this is a living document and it is intended that fuller guidance for other NMOS specifications be added in future.
```
I don't think there is an official list of `core` specifications. Should we just say:
The document focuses primarily on guidance for the following NMOS specifications:
* IS-04
* IS-05
* BCP-003-01

rmsporter replied:
Agreed. I have made that change.

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+![Role of NMOS Controller in a Networked Media System](./images/NMOS_Controller.svg)
+
+The EBU have also produced a useful diagram showing the role of Controllers within a wider context (link to be added).
```
When will the link be available?

rmsporter replied:
Not sure, so I have removed this line for now until the EBU diagram is publicly available.

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+The EBU have also produced a useful diagram showing the role of Controllers within a wider context (link to be added).
+
+This guide will cover only how the Controller interacts with the NMOS APIs. It will not include other aspects such as presentation or other features of the Client software.
```
Should we change from future tense to present?
This guide covers...

rmsporter replied:
Yes. I have made that change.

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+This guide will cover only how the Controller interacts with the NMOS APIs. It will not include other aspects such as presentation or other features of the Client software.
+
+Where a Controller is additionally acting as an Node (e.g. receiving monitoring information via IS-07), this guide will not cover any requirements relating to this.
```
Same as above.
Should we change from future tense to present?
This guide covers...

rmsporter replied:
Yes. I have made that change.

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+These, along with the provided JSON schemas and examples, often contain canonical definitions not described in the supporting text.
+
+Similarly be aware of the [NMOS Parameter Registers](https://specs.amwa.tv/nmos-parameter-registers/) as this contains constant definitions used by many of the specifications.
```
What does constant mean in this context and is it misleading?

rmsporter replied:
Yes, possibly. I have removed the word.

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+### IS-04 Discovery and Registration
+> *Controllers use the **IS-04 Query API** to discover and obtain updates on the NMOS resources that are available on the network.*
+
+The IS-04 specification describes the mechanism for the discovery and registration of NMOS resources within a media network. It comprises three APIs: the Registration API, the Query API and the Node API. Only the Query API has relevance to Controllers, but a brief summary of each API is given below for context.
```
Wouldn’t P2P controllers use mDNS and the Node API exclusively?
So maybe we need to take out `Only the Query API has relevance to Controllers`?

rmsporter replied:
Good point. I have removed that part of the sentence.

peterbrightwell replied:
Discussed on Architecture Review Group call and we're ok with removing it

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+The IS-04 specification describes the mechanism for the discovery and registration of NMOS resources within a media network. It comprises three APIs: the Registration API, the Query API and the Node API. Only the Query API has relevance to Controllers, but a brief summary of each API is given below for context.
+
+The **IS-04 Registration API** is exposed by the Registry. When a Node first joins the network, it uses [DNS-SD](https://specs.amwa.tv/info-004/) to discover the URL of the Registration API. It then uses the Registration API to register information about itself and all its sub-resources (Devices, Senders, Receivers, Sources and Flows). Each Node then communicates with the Registration API with a  ...
```
Or mDNS?

rmsporter replied:
Added.

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+The **IS-04 Registration API** is exposed by the Registry. When a Node first joins the network, it uses [DNS-SD](https://specs.amwa.tv/info-004/) to discover the URL of the Registration API. It then uses the Registration API to register information about itself and all its sub-resources (Devices, Senders, Receivers, Sources and Flows). Each Node then communicates with the Registration API with a  ...
+
+The **IS-04 Query API** is also exposed by the Registry. Controllers discover the URL of the Query API using DNS-SD and then obtain an up-to-date list of all registered Nodes and sub-resources (Devices, Senders, Receivers, Sources and Flows) so that they can indicate to the user the currently available resources on the network. Controllers also use the Query API to set up websockets subscriptions ...
```
```suggestion
The **IS-04 Query API** is also exposed by the Registry. Controllers discover the URL of the Query API using DNS-SD and then obtain an up-to-date list of all registered Nodes and sub-resources (Devices, Senders, Receivers, Sources and Flows) so that they can indicate to the user the currently available resources on the network. Controllers also use the Query API to set up WebSocket subscriptions to resources, so that they obtain updates every time something changes in the Registry.
```

rmsporter replied:
Fixed

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+The **IS-04 Query API** is also exposed by the Registry. Controllers discover the URL of the Query API using DNS-SD and then obtain an up-to-date list of all registered Nodes and sub-resources (Devices, Senders, Receivers, Sources and Flows) so that they can indicate to the user the currently available resources on the network. Controllers also use the Query API to set up websockets subscriptions ...
+
+The **IS-04 Node API** is exposed by Nodes. It is used by other Nodes to discover information about a Node directly when there is no Registry available on the network (in peer-to-peer mode).
```
It may also be used by controllers in a P2P environment.
Maybe we should say It is used by other Nodes or Controllers?

rmsporter replied:
Agreed. I have added Controllers.

cristian-recoseanu commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,100 @@
+
+The **IS-05 Connection API** is exposed by NMOS Devices. Controllers discover the URL of the Connection API through the list of `controls` for the Device. This is part of the information that will have been registered to the IS-04 Registry by the Device's parent Node. It is available to the Controller by querying the Registry using the IS-04 Query API and examining the information returned about  ...
+
+Controllers make connections between Senders and Receivers by making calls to the Connection APIs of their parent Devices. A Controller first calls the Connection API for the Sender to update its transport parameters as required and then obtain its transport file. It then calls the Connection API for the Receiver to provide it with the transport file and complete the connection.
```
If the example is meant to showcase RTP maybe we should mention that it is an RTP example, since WebSocket and MQTT have no transport file?
Otherwise, we probably need to make it more generic or cover other use cases potentially.

rmsporter replied:
Good point. I've added words to make it clear that it's an example for an RTP-based connection.

peterbrightwell replied:
Reviewed on Architecture Review Group call -- LGTU

rmsporter committed:
Update Controller Guide.md based on review comments

Updates made to Controller Guide.md based on review comments from @cristian-recoseanu

rmsporter committed:
Update Controller Guide.md

Fix WebSocket typo

peterbrightwell commented on docs/Controller Guide.md:
```diff
@@ -0,0 +1,107 @@
+
+The IS-04 specification describes the mechanism for the discovery and registration of NMOS resources within a media network. It comprises three APIs: the Registration API, the Query API and the Node API. A brief summary of each API is given below.
+
+The **IS-04 Registration API** is exposed by the Registry. When a Node first joins the network, it uses [DNS-SD](https://specs.amwa.tv/info-004/) (or mDNS) to discover the URL of the Registration API. It then uses the Registration API to register information about itself and all its sub-resources (Devices, Senders, Receivers, Sources and Flows). Each Node then communicates with the Registration A ...
```
```suggestion
The **IS-04 Registration API** is exposed by the Registry. When a Node first joins the network, it uses [unicast DNS-SD](https://specs.amwa.tv/info-004) or multicast DNS-SD (mDNS) to discover the URL of the Registration API. It then uses the Registration API to register information about itself and all its sub-resources (Devices, Senders, Receivers, Sources and Flows). Each Node then communicates with the Registration API with a regular heatbeat message to let the Registry know that it is still on the network. Nodes also use the Registration API to update the information about themselves in the Registry whenever something changes. 
```

peterbrightwell replied:
Action @AMWA-TV/nmos-architecture-review : go through NMOS docs and sanitise the use of mDNS and DNS-SD.

peterbrightwell reviewed (approved):
As the only small change ARG wanted in today's review (2022-04-12) was to do with mDNS language, we will approve and merge!

peterbrightwell committed:
Update docs/Controller Guide.md for mDNS/DNS-SD language
